### PR TITLE
[TOPI] Add C++ implementation of elementwise operators

### PR DESCRIPTION
**Pull request by nishi-t**
This PR adds c++ implementation of elementwise operators:

- elemwise_sum
- full
- full_like
- ~less~
- ~greter~

### CODE REVIEW

File: topi/include/topi/elemwise.h [hunk 1]
```diff
@@ -169,5 +171,110 @@ inline Tensor cast(const Tensor& x,
...
+/*!
+* \brief Creates an operation that sum each element of a tensor
+*
+* \param x The input tensor
+* \param num_args The number of arguments
+*
+* \return A Tensor whose op member is the sum operation
+*/
+inline Tensor elemwise_sum(const Array<Tensor>& xs,
+                           int num_args,
```
Comment (srkreddy1238) [inline topi/include/topi/elemwise.h]: What is the significance of num_args?

File: topi/include/topi/elemwise.h [hunk 2]
```diff
@@ -169,5 +171,110 @@ inline Tensor cast(const Tensor& x,
...
+* \param rhs The input tensor
+* \param out_type The type of output
+*
+* \return A Tensor filled with fill_value
+*/
+inline Tensor less(const Tensor& lhs,
+                      const Tensor& rhs,
+                      Type out_type,
+                      std::string name = "tensor",
+                      std::string tag = kElementWise) {
```
Comment (PariksheetPinjari909) [inline topi/include/topi/elemwise.h]: Indentation

File: topi/include/topi/elemwise.h [hunk 3]
```diff
@@ -169,5 +171,110 @@ inline Tensor cast(const Tensor& x,
...
+
+/*!
+* \brief Creates an operation that compare two compare input tensors element-wise
+*        and return an mask tensor which contains 1 if lhs > rhs holds else 0
+*
+* \param lhs The input tensor
+* \param rhs The input tensor
+* \param out_type The type of output
+*
+* \return A Tensor filled with fill_value
```
Comment (PariksheetPinjari909) [inline topi/include/topi/elemwise.h]: add correct doc string

File: topi/include/topi/elemwise.h [hunk 4]
```diff
@@ -169,5 +171,110 @@ inline Tensor cast(const Tensor& x,
...
+
+/*!
+* \brief Creates an operation that compare two compare input tensors element-wise
+*        and return an mask tensor which contains 1 if lhs < rhs holds else 0
+*
+* \param lhs The input tensor
+* \param rhs The input tensor
+* \param out_type The type of output
+*
+* \return A Tensor filled with fill_value
```
Comment (PariksheetPinjari909) [inline topi/include/topi/elemwise.h]: add correct doc string

File: topi/include/topi/elemwise.h [hunk 5]
```diff
@@ -169,5 +171,110 @@ inline Tensor cast(const Tensor& x,
...
+* \param fill_value The value to be filled
+*
+* \return A Tensor filled with fill_value
+*/
+inline Tensor full(const Array<Expr>& shape,
+                   Type dtype,
+                   Expr fill_value,
+                   std::string name = "tensor",
+                   std::string tag = kElementWise) {
+  Expr ev = lossless_cast(dtype, fill_value);
```
Comment (srkreddy1238) [inline topi/include/topi/elemwise.h]: It's possible that lossless_cast could return an undefined Expr.

Could raise an appropriate error here to user !

File: topi/include/topi/elemwise.h [hunk 6]
```diff
@@ -169,5 +171,110 @@ inline Tensor cast(const Tensor& x,
...
+* \param shape The input tensor
+* \param fill_value The value to be filled
+*
+* \return A Tensor filled with fill_value
+*/
+inline Tensor full_like(const Tensor& x,
+                        Expr fill_value,
+                        std::string name = "tensor",
+                        std::string tag = kElementWise) {
+  Expr ev = lossless_cast(x->dtype, fill_value);
```
Comment (srkreddy1238) [inline topi/include/topi/elemwise.h]: It's possible that lossless_cast could return an undefined Expr.

Could raise an appropriate error here to user !

File: topi/src/topi.cc [hunk 7]
```diff
@@ -174,6 +174,31 @@ TVM_REGISTER_GLOBAL("topi.cast")
   *rv = cast(args[0], args[1]);
   });
 
+TVM_REGISTER_GLOBAL("topi.elemwise_sum")
+.set_body([](TVMArgs args, TVMRetValue *rv) {
+  *rv = elemwise_sum(args[0], args[1]);
```
Comment (srkreddy1238) [inline topi/src/topi.cc]: Corresponding to above comment, number or args can be obtained inside topi by size().

File: topi/include/topi/elemwise.h (hunk 1 above)
Comment (nishi-t) [reply topi/include/topi/elemwise.h]: I add num_args to it to correspond to [python implementation](https://github.com/dmlc/tvm/blob/master/topi/python/topi/tensor.py#L8), but it is not used in there too. If not necessary, I'll remove it both cpp and python. Would you comment on this? 
\cc @tqchen

Comment (tqchen) [reply topi/include/topi/elemwise.h]: Can be remove safely

Comment (nishi-t) [reply topi/include/topi/elemwise.h]: Ok, thanks!

**Comment by nishi-t** [comment]
@PariksheetPinjari909 @srkreddy1238 Plase review again.

File: topi/include/topi/elemwise.h [hunk 8]
```diff
@@ -169,5 +171,124 @@ inline Tensor cast(const Tensor& x,
...
+      }
+      return sum_expr;
+  }, name, tag);
+}
+
+/*!
+* \brief Creates an operation that fill a tensor with fill_value
+*
+* \param shape The shape of a tensor
+* \param dtype The number of arguments
```
Comment (PariksheetPinjari909) [inline topi/include/topi/elemwise.h]: add correct doc string for dtype

File: topi/include/topi/elemwise.h [hunk 9]
```diff
@@ -169,5 +171,124 @@ inline Tensor cast(const Tensor& x,
...
+* \brief Creates an operation that compare two compare input tensors element-wise
+* and return an mask tensor which contains 1 if lhs > rhs holds else 0
+*
+* \param lhs The input tensor
+* \param rhs The input tensor
+* \param out_type The type of output
+* \param name The name of the operation
+* \param tag The tag to mark the operation
+*
+* \return A Tensor filled with fill_value
```
Comment (PariksheetPinjari909) [inline topi/include/topi/elemwise.h]: add correct doc string

File: topi/include/topi/elemwise.h [hunk 10]
```diff
@@ -169,5 +171,124 @@ inline Tensor cast(const Tensor& x,
...
+* \brief Creates an operation that compare two compare input tensors element-wise
+* and return an mask tensor which contains 1 if lhs < rhs holds else 0
+*
+* \param lhs The input tensor
+* \param rhs The input tensor
+* \param out_type The type of output
+* \param name The name of the operation
+* \param tag The tag to mark the operation
+*
+* \return A Tensor filled with fill_value
```
Comment (PariksheetPinjari909) [inline topi/include/topi/elemwise.h]: add correct doc string

File: topi/tests/python_cpp/test_topi_tensor.py [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+
+def test_comparator():
+    verify_comparator((3,4,5), "float32")
+    verify_comparator((7,), "int32")
+    verify_comparator((3,4,5), "float32", "int8")
+
+
+if __name__ == "__main__":
+    test_elemwise_sum()
```
Comment (PariksheetPinjari909) [inline topi/tests/python_cpp/test_topi_tensor.py]: call test_full and test_comparator

File: topi/include/topi/elemwise.h [hunk 12]
```diff
@@ -169,5 +171,124 @@ inline Tensor cast(const Tensor& x,
...
+*
+* \param shape The shape of a tensor
+* \param dtype The number of arguments
+* \param fill_value The value to be filled
+* \param name The name of the operation
+* \param tag The tag to mark the operation
+*
+* \return A Tensor filled with fill_value
+*/
+inline Tensor full(const Array<Expr>& shape,
```
Comment (PariksheetPinjari909) [inline topi/include/topi/elemwise.h]: why is the name full. It's operation is to fill the tensor with value fill_value. Shouldn't the name be 'fill'?
@tqchen python op also has the name 'full'. Can you clarify here.

File: topi/include/topi/elemwise.h [hunk 13]
```diff
@@ -169,5 +171,124 @@ inline Tensor cast(const Tensor& x,
...
+*
+* \param x The input tensor
+* \param fill_value The value to be filled
+* \param name The name of the operation
+* \param tag The tag to mark the operation
+*
+* \return A Tensor filled with fill_value
+*/
+inline Tensor full_like(const Tensor& x,
+                        Expr fill_value,
```
Comment (srkreddy1238) [inline topi/include/topi/elemwise.h]: const Expr fill_value

File: topi/include/topi/elemwise.h [hunk 14]
```diff
@@ -169,5 +171,124 @@ inline Tensor cast(const Tensor& x,
...
+* \param dtype The number of arguments
+* \param fill_value The value to be filled
+* \param name The name of the operation
+* \param tag The tag to mark the operation
+*
+* \return A Tensor filled with fill_value
+*/
+inline Tensor full(const Array<Expr>& shape,
+                   Type dtype,
+                   Expr fill_value,
```
Comment (srkreddy1238) [inline topi/include/topi/elemwise.h]: const Expr fill_value

File: topi/tests/python_cpp/test_topi_tensor.py [hunk 15]
```diff
@@ -0,0 +1,127 @@
+"""Test code for tensor operator"""
+import numpy as np
+import tvm
+import topi
+
+def verify_elemwise_sum(num_args, dtype):
```
Comment (srkreddy1238) [inline topi/tests/python_cpp/test_topi_tensor.py]: Remove num_args here as well.

File: topi/tests/python_cpp/test_topi_tensor.py [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+        s = topi.cpp.generic.default_schedule(target, [esum], False)
+        ctx = tvm.context(device, 0)
+        out = tvm.nd.array(np.zeros(shape, dtype=dtype), ctx)
+        f = tvm.build(s, tvm_placeholders + [esum], device, name="elemwise_sum")
+        tvm_nd = [tvm.nd.array(nd, ctx) for nd in np_nd] + [out]
+        f(*tvm_nd)
+        np_out = np.sum(np.array(np_nd), axis=0)
+        np.testing.assert_allclose(out.asnumpy(), np_out, rtol=1e-5)
+
+    for device in ["llvm"]:
```
Comment (srkreddy1238) [inline topi/tests/python_cpp/test_topi_tensor.py]: Specific reason for only llvm ?

**Comment by srkreddy1238** [comment]
@nishi-t  
other than these comments LGTM.

**Comment by nishi-t** [comment]
@srkreddy1238 @PariksheetPinjari909 thanks a lot. I'll fix it according to your comments.

**Comment by tqchen** [comment]
please rebase against master, greater and less are added in #1321 and supported as broadcast ops

File: topi/tests/python_cpp/test_topi_tensor.py (hunk 15 above)
Comment (nishi-t) [reply topi/tests/python_cpp/test_topi_tensor.py]: I use num_args here to generate test data for doing various element_wise tests.

File: topi/include/topi/elemwise.h (hunk 12 above)
Comment (nishi-t) [reply topi/include/topi/elemwise.h]: mxnet and numpy use `full` as operation name, but tensorflow use `fill`. maybe it's a matter of taste.  I'll follow community's direction. 

* https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.full.html
* https://mxnet.incubator.apache.org/api/python/symbol/symbol.html#mxnet.symbol.full
* https://www.tensorflow.org/api_docs/python/tf/fill

Comment (PariksheetPinjari909) [reply topi/include/topi/elemwise.h]: ok

**Comment by tqchen** [comment]
@nishi-t please ping the reviewers again once you have updates

**Comment by tqchen** [comment]
@PariksheetPinjari909 @srkreddy1238 can you do another round of review? please https://docs.tvm.ai/contribute/code_review.html#approve-and-request-changes-explicitly

**Comment by tqchen** [comment]
Thanks all the reviewers, this is now merged!

**Pull request merged**